plugins/GigPlayer/GigPlayer.cpp

```diff
@@ -333,89 +339,73 @@ void GigInstrument::play( SampleFrame* _working_buffer )
 
 	if( m_instance == nullptr || m_instrument == nullptr )
 	{
-		m_synthMutex.unlock();
 		m_notesMutex.unlock();
+		m_synthMutex.unlock();
 		return;
 	}
 
-	for( QList<GigNote>::iterator it = m_notes.begin(); it != m_notes.end(); ++it )
-	{
-		// Process notes in the KeyUp state, adding release samples if desired
-		if( it->state == GigState::KeyUp )
-		{
-			// If there are no samples, we're done
-			if( it->samples.empty() )
+	// TODO: C++20 std::erase_if
+	// Process and remove if complete
+	m_notes.erase(std::remove_if(m_notes.begin(), m_notes.end(),
+		[&](GigNote& gigNote) {
+			// Process notes in the KeyUp state, adding release samples if desired
+			if (gigNote.state == GigState::KeyUp)
 			{
-				it->state = GigState::Completed;
-			}
-			else
-			{
-				it->state = GigState::PlayingKeyUp;
-
-				// Notify each sample that the key has been released
-				for (auto& sample : it->samples)
+				// If there are no samples, we're done
+				if (gigNote.samples.empty()) { gigNote.state = GigState::Completed; }
+				else
 				{
-					sample.adsr.keyup();
-				}
+					gigNote.state = GigState::PlayingKeyUp;
 
-				// Add release samples if available
-				if( it->release == true )
-				{
-					addSamples( *it, true );
+					// Notify each sample that the key has been released
+					for (auto& sample : gigNote.samples) { sample.adsr.keyup(); }
+
+					// Add release samples if available
+					if (gigNote.release) { addSamples(gigNote, true); }
 				}
 			}
-		}
-		// Process notes in the KeyDown state, adding samples for the notes
-		else if( it->state == GigState::KeyDown )
-		{
-			it->state = GigState::PlayingKeyDown;
-			addSamples( *it, false );
-		}
-
-		// Delete ended samples
-		for( QList<GigSample>::iterator sample = it->samples.begin();
-				sample != it->samples.end(); ++sample )
-		{
-			// Delete if the ADSR for a sample is complete for normal
-			// notes, or if a release sample, then if we've reached
-			// the end of the sample
-			if( sample->sample == nullptr || sample->adsr.done() ||
-				( it->isRelease == true &&
-				  sample->pos >= sample->sample->SamplesTotal - 1 ) )
+			// Process notes in the KeyDown state, adding samples for the notes
+			else if (gigNote.state == GigState::KeyDown)
 			{
-				sample = it->samples.erase( sample );
-
-				if( sample == it->samples.end() )
-				{
-					break;
-				}
+				gigNote.state = GigState::PlayingKeyDown;
+				addSamples(gigNote, false);
 			}
-		}
 
-		// Delete ended notes (either in the completed state or all the samples ended)
-		if( it->state == GigState::Completed || it->samples.empty() )
+			// TODO: C++20 std::erase_if
```

Comment:
@messmerd does #7554 add this functionality?

Reply:
Yes, the member function `erase_if` will be available, and in GCC 10 the free function will also be available.